pkg/util/k8s/cluster.go

```diff
@@ -0,0 +1,42 @@
+package k8s
+
+import (
+	"context"
+
+	ocp_configv1 "github.com/openshift/api/config/v1"
+	metaerrors "k8s.io/apimachinery/pkg/api/meta"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// IsClusterBeingUpgraded checks if the Kubernetes cluster is being upgraded.
+// Currently this only checks for OpenShift clusters.
+func IsClusterBeingUpgraded(k8sClient client.Client) (bool, error) {
+	cvList := &ocp_configv1.ClusterVersionList{}
+	err := k8sClient.List(context.TODO(), cvList, &client.ListOptions{})
+	if metaerrors.IsNoMatchError(err) {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+
+	if len(cvList.Items) == 0 {
+		// Cannot detect upgrade as there is no ClusterVersion
+		return false, nil
+	}
+
+	cv := cvList.Items[0].DeepCopy()
+	if cv.Spec.DesiredUpdate == nil || cv.Spec.DesiredUpdate.Version == "" {
+		// No upgrade has been started
+		return false, nil
+	}
+
+	for _, h := range cv.Status.History {
+		if h.Version == cv.Spec.DesiredUpdate.Version {
+			return h.State != ocp_configv1.CompletedUpdate, nil
```

Comment:
I am a little concerned, openshift upgrade may be in bad state in some cases, and portworx upgrade will be stuck forever, although we have that annotation to force continue update, users may not know it.

During my past week testing, i have seen many times openshift upgrade was stuck and failed in the middle. and sometimes the failure is not related to updating master/worker node. in that case we are still ok to perform portworx upgrade.

Reply:
Yes, even I thought about that and it is a valid concern. I saw that in case of a failed ocp upgrade, if you re-trigger the upgrade to a different version, the status in history is changed again and the failed status goes away. Basically if a user has some issues with ocp upgrade then they will either retry or try to fix the upgrade.
I think conservatively we should block the portworx upgrade until they contact either the redhat support or portworx support. One case is that redhat support fixes the ocp upgrade problem, then portworx upgrade goes fine. Another case is that portworx support determines that it is safe to let portworx upgrade and asks them to continue by adding the annotation.
Ideally we want users to not even run ocp and portworx upgrades in parallel and if either one is failing and stuck, they should resolve it and then move ahead with the other one.
Anyway these are my 2 cents on this. Let's check with our support team as well on slack and see what they think about this.